cpp/include/cuvs/core/byte_array.hpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+
+namespace cuvs::detail {
+
+struct byte_array {
+  void* data     = nullptr;
+  bool is_signed = false;
```

Comment:
Passing this on gpu will probably require three words, right? This could have the register usage consequence in some edge cases. If only could we rely on the `data` being aligned - then we could use the lowest bit of it as the signedness tag and pass this struct everywhere in place of the pointers.

Reply:
The pointers will almost surely be aligned, but technically one can pass an unaligned pointer since the data is one byte granularity. My question is would it be acceptable for us to always assume it's at least two-byte aligned (or make an alignment check somewhere and throw an error otherwise)?

Reply:
I'd do something along the lines of this:
```C++
struct byte_arithmetic_ptr {
 private:
    constexpr static uintptr_t kSignMask = 0x1;
    uintptr_t value_;
  public:
    byte_arithmetic_ptr(uint8_t* ptr): value_(reinterpret_cast<uintptr_t>(ptr) {}
    byte_arithmetic_ptr(int8_t* ptr): value_(reinterpret_cast<uintptr_t>(ptr | kSignMask) {}
   
    constexpr void* get_data() {
      return reinterpret_cast<void*>(value_ & ~kSignMask);
    }
    constexpr bool is_signed() const {
      return (value_ & kSignMask) == kSignMask;
    }
};
```

Reply:
I think we can keep this in mind for later, in case performance becomes an issue? I just posted some performance #s in the description body and IMO they look pretty good :)